src/compiler/Stage/Parse/Parser.elm

```diff
@@ -410,23 +410,74 @@ literalInt =
         |> P.inContext InLiteralInt
 
 
-{-| TODO escapes
-TODO Unicode escapes
--}
+-- for literalChar and, in the future, literalString
+stringHelp = 
+    P.oneOf
+    [ P.succeed (identity)
+        |. P.token (P.Token "\\" ExpectingEscapeBackslash)
+        |= P.oneOf
+            [ P.map (\_ -> '\"') (P.token (P.Token "\"" ExpectingEscapeCharacter))
+            , P.map (\_ -> '\'') (P.token (P.Token "'"  ExpectingEscapeCharacter))
+            , P.map (\_ -> '\n') (P.token (P.Token "n"  ExpectingEscapeCharacter))
+            , P.map (\_ -> '\t') (P.token (P.Token "t"  ExpectingEscapeCharacter))
+            , P.map (\_ -> '\r') (P.token (P.Token "r"  ExpectingEscapeCharacter))
```

Comment:
I'm still thinking about these ParserProblems. The problem with this current solution is that the error messages will only be able to show five times the same thing ("I expected escape character, escape character, ...")

Please parameterize this one (`| ExpectingEscapeCharacter Char`) and give these usages of it the chars they're really about. That should be enough and not bloat the ParserProblem type so much. WDYT?

Reply:
How about something like below?

The above behavior would interpret `\x` as a backslash followed by a separate `x`.
The below would throw an error if it sees `\x`.

```diff
 |= P.oneOf
-    [ P.map (\_ -> '\"') (P.token (P.Token "\"" ExpectingEscapeCharacter))
-    , P.map (\_ -> '\'') (P.token (P.Token "'"  ExpectingEscapeCharacter))
-    , P.map (\_ -> '\n') (P.token (P.Token "n"  ExpectingEscapeCharacter))
-    , P.map (\_ -> '\t') (P.token (P.Token "t"  ExpectingEscapeCharacter))
-    , P.map (\_ -> '\r') (P.token (P.Token "r"  ExpectingEscapeCharacter))
+    [ P.map (\_ -> '\"') (Parser.token "\""))
+    , P.map (\_ -> '\'') (Parser.token "'"))
+    , P.map (\_ -> '\n') (Parser.token "n"))
+    , P.map (\_ -> '\t') (Parser.token "t"))
+    , P.map (\_ -> '\r') (Parser.token "r"))
     , P.succeed identity
         |. P.token (P.Token "u" ExpectingEscapeCharacter)
         |. P.token (P.Token "{" ExpectingUnicodeEscapeLeftBrace)
         |= unicode
         |. P.token (P.Token "}" ExpectingUnicodeEscapeRightBrace)
+    , P.problem ExpectingValidEscapeChar
     ]
```

> Please parameterize this one (`| ExpectingEscapeCharacter Char`) and give these usages of it the chars they're really about. That should be enough and not bloat the ParserProblem type so much. WDYT?

The problem is that I don't think in either solutions, both the current one in the PR and the diff above, the compiler would print out one of those token errors. Rather, it would either skip the `oneOf` (current PR code) or print `ExpectingValidEscapeChar` (diff above).

Reply:
current PR state: https://ellie-app.com/5Z9tkd2hGSQa1
my suggestion: https://ellie-app.com/5Z9vStSJv72a1
The changes are on lines 31-37, and 77
I think that's enough! The ParserProblems will tell you what characters did it expect :arrow_right: we can write a nice error message.

If you think this still overlooks some corner case, please shout! :)

Reply:
Oops, my bad!
Fixed by https://github.com/elm-in-elm/compiler/pull/20/commits/7c495f7661bc8a5641eb8b9a962872d81dce7f05